plugins/modules/source_control/gitlab/gitlab_service.py

```diff
@@ -0,0 +1,485 @@
+        return ret
+
+    @staticmethod
+    def events_equal(obj, events):
+        return all((k in obj and obj[k] == v) for k, v in events.items())
+
+    @staticmethod
+    def has_credentials(attr):
+        return ('password' in attr and attr['password']) or ('token' in attr and attr['token']) or ('api_key' in attr and attr['api_key'])
+
+    def equals(self, prev_attr, active, params, events):
+        # filtered_params = {k: v for k, v in params.items() if k not in self.credentialParams}
+        # python2.6 remove-me:
+        filtered_params = {}
+        for k, v in iteritems(params):
+            if k not in self.credentialParams:
+                filtered_params[k] = v
+
+        if prev_attr['active'] != active:
+            self._module.debug("active status differs")
+            return False
+        if self.has_credentials(params):
+            self._module.debug("has credentials: assuming difference")
+            return False
+        if not self.events_equal(prev_attr, self.expand_events(events)):
+            self._module.debug("events differs: %s != %s" % (prev_attr, events))
+            return False
+        if prev_attr['properties'] != filtered_params:
+            self._module.debug("services attributes differs: %s != %s" % (prev_attr['properties'], filtered_params))
+            return False
+        return True
+
+    def setattrs(self, obj, new_attr):
+        for k, v in new_attr.items():
+            setattr(obj, k, v)
+
+
+def main():
+    definitions = init_definitions()
+
+    base_spec = basic_auth_argument_spec()
+    base_spec.update(dict(
+        api_token=dict(type='str', no_log=True),
+        project=dict(required=True),
+        service=dict(required=True, type='str', choices=list(definitions.keys())),
+        active=dict(required=False, default=True, type='bool'),
+        params=dict(required=False, type='dict'),
+        events=dict(required=False, type='list', elements='str', default=GitLab_Services.HOOK_EVENTS, choices=GitLab_Services.HOOK_EVENTS),
+        state=dict(default='present', choices=['present', 'absent']),
+    ))
+
+    constraints = dict(
+        mutually_exclusive=[
+            ['api_username', 'api_token'],
+            ['api_password', 'api_token']
+        ],
+        required_together=[
+            ['api_username', 'api_password']
+        ],
+        required_one_of=[
+            ['api_username', 'api_token']
+        ],
+        required_if=[
+            ['state', 'present', ['params']]
+        ],
+        supports_check_mode=True
+    )
+
+    stub_init = AnsibleModule(argument_spec=base_spec, **constraints)
+    service = stub_init.params['service']
+    state = stub_init.params['state']
+    if state == 'present':
+        # since we know the service (which has been validated), recreate a module_specs to validate suboptions
+        # sub_arg_specs = {i: definitions[service][i] for i in definitions[service] if i != '_events'}
```

Comment:
This should also work with Python 2.6:
```suggestion
        sub_arg_specs = dict((i, definitions[service][i]) for i in definitions[service] if i != '_events')
```

Reply:
Nice tip! Thank you.